jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,233 @@
+
+**Note:** you need the Nim devel branch to compile the code below.
+
+## Background
+
+A channel is a model for sharing memory via message passing. A thread is able to send or receive messages over a channel. It's like sending a letter to your friend. The postman is the channel. Your friend is the receiver.  You may know  `system/channels` already exists. What’s the difference between new channels implementation and the old one? If you use the old one, you need to copy your letter by hand first and send the copied one to your friend instead. Then your friend may mark something on the copied letter, it won’t affect the original letter. It works fine, however it is not efficient. If you use the new one, you only need to put your letter in the mailbox. No need to copy your letter!
+
+## The advantages
+
+- Designed for ARC/ORC, no legacy code
+- No need to `deepcopy`, just move data around
+- No data races
+- Based on [Michael & Scott lock-based queues](https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf)
+
+## Explore the new channels
+
+**Note:** Be sure to compile your code with `--gc:orc –-threads:on`.
+
+### Let's crawl the web
+
+**todo_urls.json**
+
+```json
+{"url": ["https://google.com", "https://nim-lang.org"]}
+```
+
+**app.nim**
+
+The main thread prepares tasks via reading `todo_urls.txt`. Then it sends JSON data to a channel. The crawl thread does the crawlers’ work. It receives URL data from the channel and downloads the contents using `httpclient`.
+
+```nim
+import std/channels
+import std/[httpclient, isolation, json]
+
+
+var ch = initChan[JsonNode]() # we need to send JsonNode
+
+proc download(client: HttpClient, url: string) =
+  let response = client.get(url)
+  echo "content: "
+  echo response.body[0 .. 20]  # prints the results
+
+proc crawl =
+  var client = newHttpClient() # the crawler
+  var data: JsonNode
+  ch.recv(data) # the JSON data
+  if data != nil:
+    for url in data["url"]:
+      download(client, url.getStr)
+  client.close()
+
+proc prepareTasks(fileWithUrls: string): seq[Isolated[JsonNode]] =
+  result = @[]
+  for line in lines(fileWithUrls):
+    result.add isolate(parseJson(line)) # parse JSON file
+
+proc spawnCrawlers  =
+  var tasks = prepareTasks("todo_urls.json")
+  for t in mitems tasks: # we need a mutable view of the items"
+    ch.send move t
+
+var thr: Thread[void]
+createThread(thr, crawl) # create crawl thread
+
+spawnCrawlers()
+joinThread(thr)
+```
+
+First you need to import `std/channels`.
+
+Then you can create a channel using `initChan`. It uses `mpmc` internally which stands for multiple producer, multiple consumer. The `elements` parameter is used to specify whether a channel is buffered or not.  For unbuffered channel, the sender and the receiver block until the other side is ready. Sending data to a buffered channel blocks only when the buffer is full. Receiving data from a buffered channel blocks when the buffer is empty.
+
```

Comment:
* specify the type: `initChan` returns a `Chan[T]`

Reply:
fixed